ios/engine/KMEI/KeymanEngine/Classes/Resource Management/ResourceFileManager.swift

```diff
@@ -56,74 +56,109 @@ public class ResourceFileManager {
+      }
+    })
+  }
+
+  public func promptPackageInstall(of package: KeymanPackage, in rootVC: UIViewController) {
+    let vc = PackageInstallViewController(for: package, completionHandler: { error in
       if let err = error {
         if let kmpError = err as? KMPError {
-          self.showKMPError(kmpError)
+          let alert = self.buildKMPError(kmpError)
+          rootVC.present(alert, animated: true, completion: nil)
         }
       } else {
-        self.showSimpleAlert(title: "Success", message: "Installed successfully.")
+        let alert = self.buildSimpleAlert(title: "Success", message: "Installed successfully.")
+        rootVC.present(alert, animated: true, completion: nil)
       }
     })
 
     let nvc = UINavigationController.init(rootViewController: vc)
-    UIApplication.shared.keyWindow?.rootViewController?.present(nvc, animated: true, completion: nil)
+    rootVC.present(nvc, animated: true, completion: nil)
   }
 
-  @available(iOSApplicationExtension, unavailable)
-  public func showKMPError(_ error: KMPError) {
-    showSimpleAlert(title: "Error", message: error.rawValue)
+  public func buildKMPError(_ error: KMPError) -> UIAlertController {
+    return buildSimpleAlert(title: "Error", message: error.rawValue)
   }
 
-  @available(iOSApplicationExtension, unavailable)
-  public func showSimpleAlert(title: String, message: String) {
+  public func buildSimpleAlert(title: String, message: String) -> UIAlertController {
     let alertController = UIAlertController(title: title, message: message,
                                             preferredStyle: UIAlertController.Style.alert)
     alertController.addAction(UIAlertAction(title: "OK",
                                             style: UIAlertAction.Style.default,
                                             handler: nil))
 
-    UIApplication.shared.keyWindow?.rootViewController?.present(alertController, animated: true, completion: nil)
+    //UIApplication.shared.keyWindow?.rootViewController?.present(alertController, animated: true, completion: nil)
+    return alertController
+  }
+
+  /**
+   * Performs the actual installation of a package's resources once confirmation has been received from the user.
+   */
+  public func finalizePackageInstall(_ package: KeymanPackage, completionHandler: (Error?) -> Void) {
+    do {
+      // Time to pass the package off to the final installers - the parse__KMP methods.
+      // TODO: (14.0+) These functions should probably be refactored to within this class eventually.
+      if package.isKeyboard() {
+        try Manager.shared.parseKbdKMP(package.sourceFolder)
+      } else {
+        try Manager.parseLMKMP(package.sourceFolder)
+      }
+      completionHandler(nil)
+    } catch {
+      log.error(error as! KMPError)
+      completionHandler(error)
+    }
+
+    //this can fail gracefully and not show errors to users
+    do {
+      try FileManager.default.removeItem(at: package.sourceFolder)
```

Comment:
Does this also clean up the .kmp.zip file?

Reply:
That part, no.  Refer to this part of the previous PR:  https://github.com/keymanapp/keyman/pull/2460/files#diff-418d2541589915da6fbecd7d3cb96091R71

By copying the file `.cachesDirectory`, we let iOS know it's a temporary file.  iOS will GC that area when necessary (and possibly at other times?), removing the file eventually.  Additional details can be found [here](https://developer.apple.com/documentation/foundation/filemanager/searchpathdirectory) and [here](https://stackoverflow.com/questions/54746421/does-ios-clean-cache-directory-automatically).  As you can tell, Apple has... less than thorough documentation about the specifics.

There's also `FileManager.default.temporaryDirectory` ([link](https://developer.apple.com/documentation/foundation/filemanager/1642996-temporarydirectory)), but that's only iOS 10.0+, and we still support 9.3.5.